### DISCUSSION TITLE
BAD time TTFB from nextjs

### DISCUSSION THREAD

**Discussion by huggler** [vercel/next.js#16195] [Help]
![snapt](https://user-images.githubusercontent.com/1282674/90275718-1d0b3f80-de39-11ea-8e56-239524c948cb.png)

Hi guys,

I did all tasks for to best the performance my website, but TTFB from server is very BAD.

Anything ideia?

I using getServerSideProps in Index page for SSR

Thanks - odete.com.br

**Comment by leerob** [comment]
What external services are you trying to communicate with in `getServerSideProps`? It's possible that's where the latency is.

**Reply by Bilalkhan4086** [reply]
@leerob I am using next14 and use server actions to get the data from my custom server (in django). But TTFB is very BAD. Is there any way to improve it?

**Comment by IRediTOTO** [comment]
Do you read about getServerSideProps? Yes, it will make your site slow. Let's try use getStaticProps, supper fast.

**Comment by mattcarlotta** [comment]
Hard to pin-point where the slow-down(s) occur(s) without a repo.

But the following list will either render block or contribute to larger/more network requests:
- getInitialProps (gIP)
- getServerSideProps (gSSP)
- 3rd party CSS
- 3rd party fonts
- Unoptimized images
- Inlined images
- Service workers

## Next lifecycles

You **_might_** be able to avoid gIP/gSSP for most requests by deferring to the client. This results in a faster TTFB, but a slower TTI. I'd recommend deferring to the client because the user at least sees a spinner or a skeleton instead of a blank white page. In addition, you can use [swr](https://swr.vercel.app/) to cache the requests. Otherwise, [Incremental Static Regeneration (getStaticProps with revalidate)](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) is another option that does something similar to swr; however, it's unfortunately not a viable solution for permission-based, user-specific data.

## 3rd party resources

As of now, there's not a whole lot you can do to improve 3rd party resources. Specifically, 3rd party CSS must be imported globally, which can result in a lot of dead code. Next is trying to get around this by introducing [RFC 3rd party CSS Modules](https://github.com/vercel/next.js/discussions/15208), but I don't expect it gain much traction among the more popular UI libraries. For 3rd party fonts, you can use a CDN (like googlefonts) or utilize Next's [Static File Serving](https://nextjs.org/docs/basic-features/static-file-serving) with link `preload` and font-display `swap`ing (read more info [here](https://codeconqueror.com/blog/using-google-fonts-with-next-js)) to load fonts quickly (might see a FOUT if the font is large and/or the CDN is slow!).

## Images

With more browsers supporting better compression image formats (like [webp](https://caniuse.com/#feat=webp)), unoptimized images can also contribute to longer loading times. Thankfully, most browsers support the [picture](https://caniuse.com/#feat=picture) element which allow targeted images to be downloaded based on window size and/or browser. I noticed you're using `webp` images, so this likely isn't causing much slowdown, however, I noticed that some of your images are inlined which can result in larger HTML files. I ran some tests on a personal project and I found that the more images I allowed to be inlined, the slower the lighthouse performance (originally set at 8192b and bumped to 12288b resulted in noticable slower scores). Instead, I offloaded all images to the client by using static file serving (or alternatively, you could use a CDN).

## Service workers

Right off the bat, service workers will take at least 5-15+ points off your score because they (by design) prefetch files before they're actually needed. While I understand their purpose, to deliver a standalone app-like experience, I find them annoying to work with in development and, without an incredibly complex configuration, costly in production. I don't use them. Facebook and Amazon don't appear to use them. Sites like Trello and Codesandbox, which do use them, break when they're blocked... 😬 

I'm just not a fan forcing a user to click a button to reload the page or forcing them to close/reopen the browser to view a most-recent build. Also, most large-scale, large user-based applications create Andriod/IOS applications anyway. My ultimate take away is: Why try to simulate something that just won't ever be as good as native code? Hopefully, sometime in the near future, web-based apps will shift away from Javascript to another language with less runtime overhead (like [Rust in WebAssembly](https://youtu.be/-2BO43FvBuA?t=330)).

That said, to each their own. Use what you want.

[With service worker](https://i.imgur.com/gucBTx0.png)
[Without service worker](https://i.imgur.com/x2N6x6J.png)
